port find_uv_at_h

edoyango opened the pull request:
Small port of find_uv_at_h

edoyango committed:
port find_uv_at_h

marshallward commented:
This is probably equivalent, but could the `j` loop also be `do concurrent`, and the outer `omp target teams` directive be replaced with a `target data` region?

edoyango commented:
Last time I tried putting do concurrent around j, do concurrent couldn't privatize the array variables. Can try again with nvhpc 25.9.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Marshall Ward ***@***.***>
Sent: Wednesday, November 19, 2025 9:12:49 PM
To: marshallward/MOM6 ***@***.***>
Cc: Edward Yang ***@***.***>; Author ***@***.***>
Subject: Re: [marshallward/MOM6] port find_uv_at_h (PR #55)

[https://avatars.githubusercontent.com/u/462315?s=20&v=4]marshallward left a comment (marshallward/MOM6#55)<https://github.com/marshallward/MOM6/pull/55#issuecomment-3551890819>

This is probably equivalent, but could the j loop also be do concurrent, and the outer omp target teams directive be replaced with a target data region?

—
Reply to this email directly, view it on GitHub<https://github.com/marshallward/MOM6/pull/55#issuecomment-3551890819>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AWRE5BZ4LUESAT3SD4YTKMD35Q7CDAVCNFSM6AAAAACMQ4WEFKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKNJRHA4TAOBRHE>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

marshallward commented:
We're still running on a machine with v24.11, so no harm leaving it as-is.

marshallward reviewed (approved):
This looks like a very safe and, if necessary, easily reversible change.

marshallward commented:
I'll try to test out the OpenMP test case either today or tomorrow and see what can be done.

JorgeG94 commented on src/parameterizations/vertical/MOM_diabatic_aux.F90:
```diff
@@ -540,11 +540,10 @@ subroutine find_uv_at_h(u, v, h, u_h, v_h, G, GV, US, ea, eb, zero_mix)
-  !$OMP                                  h_neglect,ea,eb,u_h,u,v_h,v,nz)                &
-  !$OMP                          private(sum_area,Idenom,a_w,a_e,a_s,a_n,b_denom_1,b1,d1,c1)
+  !$omp target enter data map(alloc: a_w,a_e,a_s,a_n,b1,d1,c1)
+  !$omp target teams loop private(a_w,a_e,a_s,a_n,b1,d1,c1) map(to: ea, eb, h) map(from: u_h, v_h)
```
Or I wonder if the new gcc sees the target directives and tries to do something?

JorgeG94 commented on src/parameterizations/vertical/MOM_diabatic_aux.F90:
```diff
@@ -571,14 +570,14 @@ subroutine find_uv_at_h(u, v, h, u_h, v_h, G, GV, US, ea, eb, zero_mix)
         v_h(i,j,1) = (h(i,j,1)*b1(i)) * ((a_n(i)*v(i,J,1)) + (a_s(i)*v(i,J-1,1)))
       enddo
-      do k=2,nz ; do i=is,ie
+      do k=2,nz ; do concurrent (i=is:ie)
```
@marshallward I find the coupling of `do ; do concurrent` a bit weird - maybe worth to agree on a style with the gfdl people too? I don't mind the ` do ; do ; do ` but mixing both feels weird - very much a nit

marshallward replied:
Do you mean the style?  I'm treating this as a temporary style to preserve the existing indentation.  I agree it should be reformatted before going into dev/gfdl or main.

edoyango replied:
using do concurrent where possible has been functionally useful as a couple times the compiler didn't understand that the inner loop(s) are parallelizable. I've only been using plain do if the loop needs to be serial (like this one)

JorgeG94 replied:
yeah this is all style haha

marshallward committed:
Vert frict: nk_in_ml reduce locality in v

This patch adds a missing reduce(max: max_nk) locality modifier to
nk_in_ml in the v-direction.

This fixes an openmp regression that was detected in ifx 2025.2.

marshallward commented:
Looks good, and pass the soft regression (double_gyre, benchmark).
